File: src/test/java/com/github/njuro/jard/thread/ThreadIntegrationTest.java
package com.github.njuro.jard.thread;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.github.njuro.jard.attachment.AttachmentCategory;
import com.github.njuro.jard.board.BoardFacade;
import com.github.njuro.jard.board.dto.BoardDto;
import com.github.njuro.jard.board.dto.BoardForm;
import com.github.njuro.jard.board.dto.BoardSettingsDto;
import com.github.njuro.jard.common.Mappings;
import com.github.njuro.jard.common.MockRequestTest;
import com.github.njuro.jard.common.WithMockUserAuthorities;
import com.github.njuro.jard.post.PostFacade;
import com.github.njuro.jard.post.PostNotFoundException;
import com.github.njuro.jard.post.dto.PostDto;
import com.github.njuro.jard.post.dto.PostForm;
import com.github.njuro.jard.thread.dto.ThreadDto;
import com.github.njuro.jard.thread.dto.ThreadForm;
import com.github.njuro.jard.user.UserAuthority;
import java.net.URI;
import java.util.Collections;
import java.util.List;
import javax.persistence.EntityManager;
import org.apache.commons.lang3.ArrayUtils;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpMethod;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
@Transactional
class ThreadIntegrationTest extends MockRequestTest {

  private static final String API_ROOT = Mappings.API_ROOT_THREADS;

  @Autowired private EntityManager entityManager;
  @Autowired private BoardFacade boardFacade;
  @Autowired private ThreadFacade threadFacade;
  @Autowired private PostFacade postFacade;

  private BoardDto board;
  private ThreadForm threadForm;

  @BeforeEach
  void setUp() {
    board =
        boardFacade.createBoard(
            BoardForm.builder()
                .label("r")
                .name("Random")
                .boardSettingsForm(
                    BoardSettingsDto.builder()
                        .defaultPosterName("Anonymous")
                        .threadLimit(200)
                        .bumpLimit(300)
                        .attachmentCategories(Collections.singleton(AttachmentCategory.IMAGE))
                        .build())
                .build());

    threadForm =
        ThreadForm.builder()
            .subject("Test thread")
            .locked(false)
            .stickied(true)
            .postForm(
                PostForm.builder()
                    .body("Test post")
                    .name("Anonymous")
                    .password("Password")
                    .ip("127.0.0.1")
                    .sage(true)
                    .build())
            .build();
  }

  @Test
  void testCreateThread() throws Exception {
    performMockMultipartRequest(
            HttpMethod.PUT,
            API_ROOT,
            buildMultipartParam("threadForm", threadForm),
            buildMultipartFile("attachment", "test_attachment_1.png"))
        .andExpect(status().isOk())
        .andExpect(nonEmptyBody());

    assertThat(threadFacade.resolveThread(board.getLabel(), 1L)).isNotNull();
  }

  @Test
  void testReplyToThread() throws Exception {
    var thread = createThreadDirectly();

    PostForm reply = PostForm.builder().body("Reply").build();

    performMockMultipartRequest(
            HttpMethod.PUT,
            buildUri(API_ROOT + Mappings.PATH_VARIABLE_THREAD, thread.getThreadNumber()),
            buildMultipartParam("postForm", reply),
            buildMultipartFile("attachment", "test_attachment_2.png"))
        .andExpect(status().isOk())
        .andExpect(nonEmptyBody());

    assertThat(threadFacade.getThread(thread).getReplies()).hasSize(1);
  }

  @Test
  void testGetThread() throws Exception {
    var thread = createThreadDirectly();

    var result =
        performMockRequest(
                HttpMethod.GET,
                buildUri(API_ROOT + Mappings.PATH_VARIABLE_THREAD, thread.getThreadNumber()))
            .andExpect(status().isOk())
            .andExpect(nonEmptyBody())
            .andReturn();

    assertThat(getResponse(result, ThreadDto.class).getThreadNumber())
        .isEqualTo(thread.getThreadNumber());
  }

  @Test
  void testGetNewReplies() throws Exception {
    var thread = createThreadDirectly();
    var reply1 = createReplyDirectly(thread);
    var reply2 = createReplyDirectly(thread);
    var reply3 = createReplyDirectly(thread);

    var result =
        performMockRequest(
                HttpMethod.GET,
                buildUri(
                    API_ROOT
                        + Mappings.PATH_VARIABLE_THREAD
                        + "/new-replies?lastPost="
                        + reply1.getPostNumber(),
                    thread.getThreadNumber()))
            .andExpect(status().isOk())
            .andExpect(nonEmptyBody())
            .andReturn();

    assertThat(getResponseCollection(result, List.class, PostDto.class))
        .extracting(PostDto::getPostNumber)
        .containsExactly(reply2.getPostNumber(), reply3.getPostNumber());
  }

  @Test
  @WithMockUserAuthorities(UserAuthority.TOGGLE_STICKY_THREAD)
  void testToggleStickyOnThread() throws Exception {
    var thread = createThreadDirectly();

    boolean originalSticky = threadForm.isStickied();

    performMockRequest(
            HttpMethod.POST,
            buildUri(
                API_ROOT + Mappings.PATH_VARIABLE_THREAD + "/sticky", thread.getThreadNumber()))
        .andExpect(status().isOk());

    assertThat(threadFacade.resolveThread(board.getLabel(), thread.getThreadNumber()).isStickied())
        .isNotEqualTo(originalSticky);
  }

  @Test
  @WithMockUserAuthorities(UserAuthority.TOGGLE_LOCK_THREAD)
  void testToggleLockOnThread() throws Exception {
    var thread = createThreadDirectly();

    boolean originalLock = thread.isLocked();

    performMockRequest(
            HttpMethod.POST,
            buildUri(API_ROOT + Mappings.PATH_VARIABLE_THREAD + "/lock", thread.getThreadNumber()))
        .andExpect(status().isOk());

    assertThat(threadFacade.resolveThread(board.getLabel(), thread.getThreadNumber()).isLocked())
        .isNotEqualTo(originalLock);
  }

  @Test
  @WithMockUserAuthorities(UserAuthority.DELETE_POST)
  void testDeleteThread() throws Exception {
    var thread = createThreadDirectly();

    assertThat(postFacade.resolvePost(board.getLabel(), thread.getOriginalPost().getPostNumber()))
        .isNotNull();

    entityManager.clear();

    performMockRequest(
            HttpMethod.DELETE,
            buildUri(
                API_ROOT + Mappings.PATH_VARIABLE_THREAD + "/" + Mappings.PATH_VARIABLE_POST,
                thread.getThreadNumber(),
                thread.getOriginalPost().getPostNumber()))
        .andExpect(status().isOk());

    assertThatThrownBy(
            () ->
                postFacade.resolvePost(board.getLabel(), thread.getOriginalPost().getPostNumber()))
        .isInstanceOf(PostNotFoundException.class);
    assertThatThrownBy(() -> threadFacade.resolveThread(board.getLabel(), thread.getThreadNumber()))
        .isInstanceOf(ThreadNotFoundException.class);
  }

  @Test
  @WithMockUserAuthorities(UserAuthority.DELETE_POST)
  void testDeletePost() throws Exception {
    var thread = createThreadDirectly();
    var reply = createReplyDirectly(thread);

    assertThat(postFacade.resolvePost(board.getLabel(), reply.getPostNumber())).isNotNull();
    assertThat(threadFacade.getThread(thread).getReplies()).hasSize(1);

    performMockRequest(
            HttpMethod.DELETE,
            buildUri(
                API_ROOT + Mappings.PATH_VARIABLE_THREAD + "/" + Mappings.PATH_VARIABLE_POST,
                thread.getThreadNumber(),
                reply.getPostNumber()))
        .andExpect(status().isOk());

    assertThatThrownBy(() -> postFacade.resolvePost(board.getLabel(), reply.getPostNumber()))
        .isInstanceOf(PostNotFoundException.class);
    assertThat(threadFacade.getThread(thread).getReplies()).isEmpty();
  }

  @Override
  protected URI buildUri(String url, Object... pathVariables) {
    return super.buildUri(url, ArrayUtils.addFirst(pathVariables, board.getLabel()));
  }

  private ThreadDto createThreadDirectly() throws Exception {
    threadForm
        .getPostForm()
        .setAttachment(buildMultipartFile("attachment", "test_attachment_1.png"));
    return threadFacade.createThread(threadForm, board);
  }

  private PostDto createReplyDirectly(ThreadDto thread) throws Exception {
    return threadFacade.replyToThread(
        PostForm.builder().body("Reply").ip("127.0.0.1").build(), thread);
  }
}

File: src/test/kotlin/com/github/njuro/jard/thread/ThreadIntegrationTest.kt
package com.github.njuro.jard.thread

import com.github.njuro.jard.MockMvcTest
import com.github.njuro.jard.TEST_ATTACHMENT_PNG
import com.github.njuro.jard.TestDataRepository
import com.github.njuro.jard.WithContainerDatabase
import com.github.njuro.jard.WithMockJardUser
import com.github.njuro.jard.attachment.AttachmentCategory
import com.github.njuro.jard.board
import com.github.njuro.jard.board.Board
import com.github.njuro.jard.boardSettings
import com.github.njuro.jard.common.InputConstraints.MAX_NAME_LENGTH
import com.github.njuro.jard.common.InputConstraints.MAX_SUBJECT_LENGTH
import com.github.njuro.jard.common.Mappings
import com.github.njuro.jard.multipartFile
import com.github.njuro.jard.post
import com.github.njuro.jard.post.dto.DeleteOwnPostDto
import com.github.njuro.jard.post.dto.PostDto
import com.github.njuro.jard.post.dto.PostForm
import com.github.njuro.jard.randomString
import com.github.njuro.jard.thread
import com.github.njuro.jard.thread.dto.ThreadDto
import com.github.njuro.jard.thread.dto.ThreadForm
import com.github.njuro.jard.toForm
import com.github.njuro.jard.user.UserAuthority
import io.kotest.matchers.booleans.shouldBeFalse
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.optional.shouldBeEmpty
import io.kotest.matchers.optional.shouldBePresent
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.mock.web.MockMultipartFile
import org.springframework.test.web.servlet.delete
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.multipart
import org.springframework.test.web.servlet.patch
import org.springframework.transaction.annotation.Transactional

@WithContainerDatabase
@Transactional
internal class ThreadIntegrationTest : MockMvcTest() {

    @Autowired
    private lateinit var db: TestDataRepository

    private lateinit var board: Board

    @BeforeEach
    fun createBoard() {
        board = db.insert(
            board(
                label = "r",
                settings = boardSettings(attachmentCategories = mutableSetOf(AttachmentCategory.IMAGE))
            )
        )
    }

    @Nested
    @DisplayName("create thread")
    inner class CreateThread {
        private fun createThread(
            threadForm: ThreadForm,
            attachment: MockMultipartFile? = multipartFile("attachment", TEST_ATTACHMENT_PNG)
        ) =
            mockMvc.multipart(Mappings.API_ROOT_THREADS, board.label) {
                part("threadForm", threadForm)
                if (attachment != null) file(attachment)
            }

        @Test
        fun `create valid thread`() {
            val threadForm = thread(board, subject = "Test subject").toForm()
            createThread(threadForm).andExpect { status { isCreated() } }
                .andReturnConverted<ThreadDto>().threadNumber shouldBe 1L
        }

        @Test
        fun `don't create invalid thread`() {
            val threadForm = thread(board, subject = randomString(MAX_SUBJECT_LENGTH + 1)).toForm()
            createThread(threadForm).andExpect { status { isBadRequest() } }
        }
    }

    @Nested
    @DisplayName("reply to thread")
    inner class ReplyToThread {
        private fun replyToThread(
            postForm: PostForm,
            thread: Thread,
            attachment: MockMultipartFile? = multipartFile("attachment", TEST_ATTACHMENT_PNG)
        ) =
            mockMvc.multipart(
                "${Mappings.API_ROOT_THREADS}/${Mappings.PATH_VARIABLE_THREAD}",
                board.label,
                thread.threadNumber
            ) {
                part("postForm", postForm)
                if (attachment != null) file(attachment)
            }

        @Test
        fun `create valid reply`() {
            val thread = db.insert(thread(board))
            val postForm = post(thread).toForm()

            replyToThread(postForm, thread).andExpect { status { isCreated() } }
                .andReturnConverted<PostDto>().postNumber shouldBe 1L
        }

        @Test
        fun `don't create invalid reply`() {
            val thread = db.insert(thread(board))
            val postForm = post(thread, name = randomString(MAX_NAME_LENGTH + 1)).toForm()

            replyToThread(postForm, thread).andExpect { status { isBadRequest() } }
        }
    }

    @Nested
    @DisplayName("get thread")
    inner class GetThread {
        private fun getThread(threadNumber: Long) = mockMvc.get(
            "${Mappings.API_ROOT_THREADS}/${Mappings.PATH_VARIABLE_THREAD}",
            board.label,
            threadNumber
        ) { setUp() }

        @Test
        fun `get existing thread`() {
            val thread = db.insert(thread(board))
            getThread(thread.threadNumber).andExpect { status { isOk() } }
                .andReturnConverted<ThreadDto>().threadNumber shouldBe 1L
        }

        @Test
        fun `don't get non-existing thread`() {
            getThread(0L).andExpect { status { isNotFound() } }
        }
    }

    @Test
    fun `get new replies for thread`() {
        val thread = db.insert(thread(board))
        db.insert(post(thread, postNumber = 2L))
        db.insert(post(thread, postNumber = 3L))

        mockMvc.get(
            "${Mappings.API_ROOT_THREADS}/${Mappings.PATH_VARIABLE_THREAD}/new-replies?lastPost=1",
            board.label,
            thread.threadNumber
        ) { setUp() }.andExpect { status { isOk() } }.andReturnConverted<List<PostDto>>() shouldHaveSize 2
    }

    @Test
    @WithMockJardUser(UserAuthority.TOGGLE_STICKY_THREAD)
    fun `toggle sticky on thread`() {
        val thread = db.insert(thread(board))

        thread.isStickied.shouldBeFalse()
        mockMvc.patch(
            "${Mappings.API_ROOT_THREADS}/${Mappings.PATH_VARIABLE_THREAD}/sticky",
            board.label,
            thread.threadNumber
        ) { setUp() }.andExpect { status { isOk() } }
        db.select(thread).shouldBePresent { it.isStickied.shouldBeTrue() }
    }

    @Test
    @WithMockJardUser(UserAuthority.TOGGLE_LOCK_THREAD)
    fun `toggle lock on thread`() {
        val thread = db.insert(thread(board))

        thread.isLocked.shouldBeFalse()
        mockMvc.patch(
            "${Mappings.API_ROOT_THREADS}/${Mappings.PATH_VARIABLE_THREAD}/lock",
            board.label,
            thread.threadNumber
        ) { setUp() }.andExpect { status { isOk() } }
        db.select(thread).shouldBePresent { it.isLocked.shouldBeTrue() }
    }

    @Nested
    @DisplayName("delete post")
    @WithMockJardUser(UserAuthority.DELETE_POST)
    inner class DeletePost {
        private fun deletePost(postNumber: Long, threadNumber: Long) = mockMvc.delete(
            "${Mappings.API_ROOT_THREADS}/${Mappings.PATH_VARIABLE_THREAD}/${Mappings.PATH_VARIABLE_POST}",
            board.label,
            threadNumber,
            postNumber
        ) { setUp() }

        @Test
        fun `delete post`() {
            val thread = db.insert(thread(board))

            deletePost(thread.originalPost.postNumber, thread.threadNumber).andExpect { status { isOk() } }
            db.select(thread).shouldBeEmpty()
        }
    }

    @Nested
    @DisplayName("delete own post")
    inner class DeleteOwnPost {
        private fun deleteOwnPost(postNumber: Long, threadNumber: Long, deletionCode: String) = mockMvc.delete(
            "${Mappings.API_ROOT_THREADS}/${Mappings.PATH_VARIABLE_THREAD}/${Mappings.PATH_VARIABLE_POST}/delete-own",
            board.label,
            threadNumber,
            postNumber
        ) { body(DeleteOwnPostDto(deletionCode)) }

        @Test
        fun `delete own post`() {
            val thread = db.insert(thread(board))
            val reply = db.insert(post(thread, deletionCode = "abcde", postNumber = 2L))

            db.select(reply).shouldBePresent()
            deleteOwnPost(reply.postNumber, thread.threadNumber, "abcde").andExpect { status { isOk() } }
            db.select(reply).shouldBeEmpty()
        }

        @Test
        fun `don't delete own post with incorrect deletion code`() {
            val thread = db.insert(thread(board))
            val reply = db.insert(post(thread, deletionCode = "abcde", postNumber = 2L))

            db.select(reply).shouldBePresent()
            deleteOwnPost(reply.postNumber, thread.threadNumber, "ghfij").andExpect { status { isBadRequest() } }
            db.select(reply).shouldBePresent()
        }
    }
}
